### PR TITLE
Add middleware to verify JWTs on API requests

### PR DESCRIPTION
**Pull request by marijnvanderhorst**
## Proposed changes
See title. This can be used by applications using this package to easily protect API endpoints by verifying the Bearer tokens with Auth0. The middleware is heavily inspired by [Auth0's guide](https://auth0.com/docs/quickstart/backend/laravel/01-authorization), albeit better and fully tested.

## Types of changes

What types of changes does your code introduce to this repository?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Testing
Explain here how to run your unit tests, and explain how to execute manual testing. 

### Unit testing
As always.

### Manual testing
Use package in laravel app and protect an API endpoint with the JWT middleware.

## Further comments
n/a

### CODE REVIEW

File: src/MrdAuth0LaravelServiceProvider.php [hunk 1]
```diff
@@ -31,10 +38,24 @@ public function register()
...
             return new Repository\Auth0UserRepository($config['model'], $config['jwt-model']);
         });
+
+        // Add a singleton for the Authentication SDK using laravel-auth0 config values to the service container.
+        $this->app->singleton(Authentication::class, function () {
+            $config = config('laravel-auth0');
+            return new Authentication(
+                $config['domain'],
+                $config['client_id'],
+                $config['client_secret'] ?? '',
```
Comment (EmielSchmeink) [inline src/MrdAuth0LaravelServiceProvider.php:54]: Ternary?

File: tests/Feature/JwtAuthorizationTest.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        $this->mockAuth0Service(null);
+
+        $this->request(false)
+            ->assertUnauthorized()
+            ->assertSee('Bearer token missing');
+    }
+
+    /**
+     * Verify the request is unauthorized when the bearer token is invalid.
```
Comment (EmielSchmeink) [inline tests/Feature/JwtAuthorizationTest.php:70]: And what if it is not null but still invalid (not adhering to the jwt standard)?

**Review by EmielSchmeink — COMMENTED** [review]
I had a quick look, and it seems good!

File: src/MrdAuth0LaravelServiceProvider.php (hunk 1 above)
Comment (marijnvanderhorst) [reply src/MrdAuth0LaravelServiceProvider.php:54]: It's actually PHP's [Null coalescing operator](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op)

File: tests/Feature/JwtAuthorizationTest.php (hunk 2 above)
Comment (marijnvanderhorst) [reply tests/Feature/JwtAuthorizationTest.php:70]: Hahaha I believe you asked the exact same question when I implemented this test in the psmanager. The underlying jwt library determines whether a token is invalid. However, that library is mocked out. By setting the jwt to `null` in this case we do not actually send a `null` token, but rather immediately trigger the invalidTokenException which the library would throw as well. Thus, the only assumption we are making here is that the underlying library is correct and throws the correct exception (which is the best we can do anyway).

File: tests/Feature/JwtAuthorizationTest.php [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+{
+    private const ROUTE_URI = 'test';
+
+    /**
+     * Define environment setup.
+     *
+     * @param  Application  $app
+     * @return void
+     */
+    protected function getEnvironmentSetUp($app)
```
Comment (SquarePlayn) [inline tests/Feature/JwtAuthorizationTest.php:23]:
```suggestion
    protected function setEnvironmentSetUp($app)
```

Comment (SquarePlayn) [reply tests/Feature/JwtAuthorizationTest.php:23]: This function doesn't return anything. Do you mean set i.o. get? Or perhaps just setupEnvironment

File: tests/Feature/JwtAuthorizationTest.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        $app['config']->set('laravel-auth0', [
+            'domain'     => 'auth.marketredesign.com',
+            'client_id'  => '123',
+        ]);
+    }
+
+    /**
+     * Perform a GET request to the test endpoint, optionally including a Bearer token in the Authorization header.
+     *
+     * @param bool $includeBearer Whether or not to include a bearer token in the request.
```
Comment (SquarePlayn) [inline tests/Feature/JwtAuthorizationTest.php:35]: $scope is missing

File: tests/Feature/JwtAuthorizationTest.php [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+     * @param Closure|null $responseHandler Response handler for endpoint. When null, a JSON containing `test` will be
+     * returned.
+     * @return TestResponse
+     */
+    private function request(bool $includeBearer, string $scope = '', ?Closure $responseHandler = null)
+    {
+        // Define a very simply testing endpoint, protected by the jwt middleware.
```
Comment (SquarePlayn) [inline tests/Feature/JwtAuthorizationTest.php:42]:
```suggestion
        // Define a very simple testing endpoint, protected by the jwt middleware.
```

File: tests/Feature/JwtAuthorizationTest.php [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    public function testScopeRequiredOneScopeProvided()
+    {
+        $this->mockAuth0Service(['scope' => 'test']);
+
+        $this->request(true, 'test')
+            ->assertOk()
+            ->assertSee('test');
+    }
+
```
Comment (SquarePlayn) [inline tests/Feature/JwtAuthorizationTest.php:126]: Suggestion: test no access if a scope is provided but it's not the correct one

File: tests/Feature/JwtAuthorizationTest.php [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+use Closure;
+use Illuminate\Foundation\Application;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Route;
+use Illuminate\Testing\TestResponse;
+use Marketredesign\MrdAuth0Laravel\Tests\TestCase;
+
+class JwtAuthorizationTest extends TestCase
+{
+    private const ROUTE_URI = 'test';
```
Comment (SquarePlayn) [inline tests/Feature/JwtAuthorizationTest.php:15]: The userID, route_URI and response of the testing-route are all 'test', making it confusing to distinguish plus can go wrong if they return the wrong one which also happens to be 'test'. I suggest using 'test_route' and such.

File: tests/Unit/CheckJwtMiddlewareTest.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+class CheckJwtMiddlewareTest extends TestCase
+{
+    /**
+     * Define environment setup.
+     *
+     * @param  Application  $app
+     * @return void
+     */
+    protected function getEnvironmentSetUp($app)
```
Comment (SquarePlayn) [inline tests/Unit/CheckJwtMiddlewareTest.php:21]: Same as in other test, this doesn't return anything so not a getter

Comment (SquarePlayn) [reply tests/Unit/CheckJwtMiddlewareTest.php:21]: Oh perhaps you mean it as like "Hey, can you get the environment set up?", so not a getter but get as in getting/doing/making? Either way, confusing.

File: tests/Feature/JwtAuthorizationTest.php (hunk 2 above)
Comment (EmielSchmeink) [reply tests/Feature/JwtAuthorizationTest.php:70]: Oops. In my defense I'm very busy.

File: src/MrdAuth0LaravelServiceProvider.php (hunk 1 above)
Comment (EmielSchmeink) [reply src/MrdAuth0LaravelServiceProvider.php:54]: TIL

**Review by EmielSchmeink — DISMISSED** [review]

**Review by SquarePlayn — CHANGES_REQUESTED** [review]
I've reviewed the code but still have to perform manual testing.

File: tests/Feature/JwtAuthorizationTest.php (hunk 3 above)
Comment (marijnvanderhorst) [reply tests/Feature/JwtAuthorizationTest.php:23]: This overwrites a function in the parent (Orchestra\Testbench\TestCase), so no way to change the name.

File: tests/Feature/JwtAuthorizationTest.php (hunk 6 above)
Comment (marijnvanderhorst) [reply tests/Feature/JwtAuthorizationTest.php:126]: This should be tested by `testScopeRequiredIncorrectScopeProvided` around line 107.

File: tests/Unit/CheckJwtMiddlewareTest.php (hunk 8 above)
Comment (marijnvanderhorst) [reply tests/Unit/CheckJwtMiddlewareTest.php:21]: I fully agree, i have no idea what there were thinking when creating this method. Unfotunately nothing we can do about it though :(

File: tests/Feature/JwtAuthorizationTest.php (hunk 7 above)
Comment (marijnvanderhorst) [reply tests/Feature/JwtAuthorizationTest.php]: Updated as suggested

File: tests/Feature/JwtAuthorizationTest.php (hunk 4 above)
Comment (marijnvanderhorst) [reply tests/Feature/JwtAuthorizationTest.php:35]: Added

File: tests/TestCase.php [hunk 9]
```diff
@@ -3,15 +3,52 @@
...
-
+use Auth0\Login\Auth0Service;
+use Auth0\SDK\API\Authentication;
+use Auth0\SDK\Exception\InvalidTokenException;
+use Illuminate\Support\Arr;
 use Marketredesign\MrdAuth0Laravel\MrdAuth0LaravelServiceProvider;
 
 class TestCase extends \Orchestra\Testbench\TestCase
 {
+    protected $userId = 'test';
```
Comment (SquarePlayn) [inline tests/TestCase.php:14]:
```suggestion
    protected $userId = 'test_user_id';
```

File: tests/Unit/CheckJwtMiddlewareTest.php [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+        // Make sure a bearer token is set.
+        $request->headers->set('Authorization', 'Bearer token');
+
+        $nextCalled = false;
+
+        // Require 'test' scope to be present, which is present in the JWT.
+        $middleware->handle($request, function () use (&$nextCalled) {
+            $nextCalled = true;
+        }, 'test');
```
Comment (SquarePlayn) [inline tests/Unit/CheckJwtMiddlewareTest.php:190]: This test file also contains quite some 'test'. I think they are all scopes, but it'd be great if you can change them all to contain some non-'test' value.

**Review by SquarePlayn — APPROVED** [review]

**Review by EmielSchmeink — APPROVED** [review]

**Pull request merged**